base/system/winlogon/sound.c

```diff
@@ -0,0 +1,325 @@
+
+            if (uNumDevs == 0)
+            {
+                if (!bLogon)
+                {
+                    Beep(440, 125);
+                }
+
+                FreeLibrary(hLibrary);
+
+                return FALSE;
+            }
+        }
+
+        _PlaySoundW = (PFPLAYSOUNDW)GetProcAddress(hLibrary, "PlaySoundW");
+
+        if (_PlaySoundW)
+        {
+            bRet = _PlaySoundW(lpFileName, NULL, dwFlags);
+        }
+
+        FreeLibrary(hLibrary);
+    }
+
+    return bRet;
+}
+
+static
+DWORD
+WINAPI
+PlaySystemSoundThread(
+    IN LPVOID lpParameter)
+{
+    PWL_PLAYSOUND_DATA PSData;
+    BOOL bLogon;
+    LPWSTR pszSndKey;
+    HKEY hUserKey, hSndKey;
+    DWORD dwType, dwSize;
+    LONG lError;
+    LPWSTR pszValue = NULL;
+    LPWSTR pszSndPath = NULL;
+
+    PSData = (PWL_PLAYSOUND_DATA)lpParameter;
+    bLogon = (PSData->Sound == SYSTEMSND_LOGON);
+
+    if (bLogon)
+    {
+        SERVICE_STATUS_PROCESS Info;
+        DWORD dwBytesNeeded;
+        ULONG Index = 0;
+        SC_HANDLE hSCManager, hService;
+
+        /* Open the service manager */
+        hSCManager = OpenSCManagerW(NULL, NULL, SC_MANAGER_CONNECT);
+
+        if (!hSCManager)
+        {
+            ERR("OpenSCManagerW failed (%x)\n", GetLastError());
+            goto Cleanup;
+        }
+
+        /* Open the wdmaud service */
+        hService = OpenServiceW(hSCManager, L"wdmaud", GENERIC_READ);
+
+        if (!hService)
+        {
+            /* The service is not installed */
+            TRACE("Failed to open wdmaud service (%x)\n", GetLastError());
+            CloseServiceHandle(hSCManager);
+            goto Cleanup;
+        }
+
+        /* Wait for wdmaud to start */
+        do
+        {
+            if (!QueryServiceStatusEx(hService, SC_STATUS_PROCESS_INFO, (LPBYTE)&Info, sizeof(Info), &dwBytesNeeded))
+            {
+                TRACE("QueryServiceStatusEx failed (%x)\n", GetLastError());
+                break;
+            }
+
+            if (Info.dwCurrentState == SERVICE_RUNNING)
+            {
+                break;
+            }
+
+            Sleep(1000);
+
+        } while (Index++ < 20);
+
+        CloseServiceHandle(hService);
+        CloseServiceHandle(hSCManager);
+
+        /* If wdmaud is not running exit */
+        if (Info.dwCurrentState != SERVICE_RUNNING)
+        {
+            WARN("wdmaud has not started!\n");
+            goto Cleanup;
+        }
+
+        /* Sound subsystem is running. Play logon sound. */
+    }
+
+    /* Impersonate current user */
+    if (!ImpersonateLoggedOnUser(PSData->Session->UserToken))
+    {
+        goto Cleanup;
+    }
+
+    /* Open user's HKCU */
+    lError = RegOpenCurrentUser(KEY_READ, &hUserKey);
+
+    RevertToSelf();
+
+    if (lError != ERROR_SUCCESS)
+    {
+        goto Cleanup;
+    }
+
+    switch (PSData->Sound)
+    {
+        case SYSTEMSND_LOGON:
+            pszSndKey = L"AppEvents\\Schemes\\Apps\\.Default\\WindowsLogon\\.Current";
```

Comment:
This is PlaySoundW's job to do this mapping (Windows' winlogon doesn't contain such strings)

Reply:
It has to do with user impersonation stuff so calling `PlaySoundW` with default behavior won't work in this case because `PlaySoundW` reads registry from HKEY_CURRENT_USER (predefined key).

Reply:
Shouldn't the PlaySoundW be called *within* an `ImpersonateLoggedOnUser` / `RevertToSelf` call pair in that case?

Reply:
That doesn't work because

> The HKEY_CURRENT_USER key maps to the root of the current user's branch in the HKEY_USERS key. It is cached for all threads in a process. **Therefore, this value does not change when another user's profile is loaded**. RegOpenCurrentUser uses the thread's token to access the appropriate key, or the default if the profile is not loaded.

[https://docs.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regopencurrentuser](https://docs.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regopencurrentuser)